brontide/noise.go

```diff
@@ -743,7 +748,19 @@ func (b *Machine) WriteMessage(w io.Writer, p []byte) error {
 // ReadMessage attempts to read the next message from the passed io.Reader. In
 // the case of an authentication error, a non-nil error is returned.
 func (b *Machine) ReadMessage(r io.Reader) ([]byte, error) {
-	_, err := io.ReadFull(r, b.nextCipherHeader[:])
+	conn, ok := r.(net.Conn)
```

Comment:
This leaks the concrete implementation of the `io.Reader`. What's the rationale of pushing the deadline iteration down into the brontide layer rather than keeping it one step above at the peer?

Reply:
the abstraction is already somewhat broken imo bc it doesn't distinguish between adversarial vs non-adversarial `io.Reader` instances, e.g. connections vs files. I considered adding a separate `ReadConn` method to handle this

setting the timeout at the peer level will likely cause reads to timeout while the read is being scheduled, since it is unaware of when the task is being scheduled by the pool.

there may also be anan arbitrary amount of time between reading payload lengths (which we want to block for), though the header is received we expect the body of the message to follow shortly after. this was also less ugly than having to have something like this in the peer:
```golang
retry:
conn.SetReadDeadline(deadline)
_, err := conn.ReadMessage()
if nerr := err.(net.Error); nerr.Timeout() {
     goto retry
}
```
as this would mean the read timeout will be triggered multiple times if the peer doesn't send us anything for a minute and we want the read timeout to be something short like 5s

Reply:
another consequence of setting the read deadline at the peer level is that we may successfully read the message header, but then timeout before/while reading the body. in this case, and w/ the retry logic, we will try again and start reading a body as if it were the header since we aren't distinguishing timeouts on header vs body

Reply:
we can push the read timeout abstraction up a bit by splitting `ReadMessage` into `ReadHeader` and `ReadBody`. this way the caller can block arbitrarily long on headers, and then set a real deadline for bodies. because of the scheduling tho, it would still be more prone to timing out while being scheduled than if the timeout is set once the task is scheduled

I think we could get around this a little more by having `ReadMessage` accept the buffer we want to read into, then from the peer we can do:
```
conn.SetReadDeadline(time.Time{})
conn.ReadHeader()
readPool.Submit(func(b *buffer.Read) error {
    conn.SetReadDeadline(time.Now().Add(readTimeout)
    rawBytes, err := conn.ReadNextBody(b) // internally calls Noise.ReadMessage(reader)
    return err
})
```
how does that sound?

Reply:
i think this also allows us to keep `ReadMessage` and note that it should only be used for non-adverserial readers, and the caller needs more control they should use the split header/body interface. interested in hearing your thoughts

Reply:
Nice, I think that last fragment gives us the best of both worlds. Also nice to patch the issue that exists right now that allows a peer to block us by writing the brontide header and nothing else. 

Re the first comment, I don't think it's the case that the abstraction is broken if it doesn't distinguish between sockets, files, etc. On the contrary, that's the point of having something like an `io.Reader` in the first place: we don't care what the underlying implementation _is_.